scope.go

```diff
@@ -261,118 +393,206 @@ func (s *scope) reportLoopRun() {
 
 // reports current registry with scope status lock held
 func (s *scope) reportRegistryWithLock() {
-	s.registry.RLock()
+	s.registry.Lock()
 	if s.reporter != nil {
-		for _, ss := range s.registry.subscopes {
-			ss.report(s.reporter)
+		for name, ss := range s.registry.subscopes {
+			if expired := ss.report(s.reporter); expired {
+				delete(s.registry.subscopes, name)
+				atomic.StoreUint32(&ss.expired, 1)
+			}
 		}
 	} else if s.cachedReporter != nil {
-		for _, ss := range s.registry.subscopes {
-			ss.cachedReport(s.cachedReporter)
+		for name, ss := range s.registry.subscopes {
+			if expired := ss.cachedReport(s.cachedReporter); expired {
+				delete(s.registry.subscopes, name)
+				atomic.StoreUint32(&ss.expired, 1)
+			}
 		}
 	}
-	s.registry.RUnlock()
+	s.registry.Unlock()
+}
+
+// getExpirySeconds gets the expiry period of metrics for this scope
+func (s *scope) getExpirySeconds() int64 {
+	return s.registry.expirePeriodSeconds
+}
+
+// hasExpired indicates whether the scope is expired
+func (s *scope) hasExpired() bool {
+	return atomic.LoadUint32(&s.expired) == 1
+}
+
+// getCurrentScope gets the current active scope which could
+// be this scope or another if another was created after this
+// one expired
+func (s *scope) getCurrentScope() *scope {
+	if s.hasExpired() {
+		// scope has expired. Go grab the current one from the registry
+		// TODO(martinm) - cache the registryKey in scope if expired occurs often
+		registryName := scopeRegistryKey(s.prefix, s.tags)
+		s.registry.RLock()
+		curScope, exists := s.registry.subscopes[registryName]
+		s.registry.RUnlock()
+
+		if exists {
```

Comment:
is it possible the scope we just got out has also expired?

Reply:
The logic to expire scopes happens within a s.registry.Lock() so when we look it up in line 434, it should exist. It could of course expire right after the RUnlock(), but that's why we do a final check before we return a metric from a counter/gauge/timer call to ensure the scope hasn't expired. If it has, we call the same function again to get a fresh scope.